File: leetcode/38-Count-And_Say/answer.c
// https://leetcode.com/problems/count-and-say/

#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#define MAX_LEN 4096

char* next_num(char* num) {

  int count = 1;
  int index = 0, next_index = 0;

  int length = strlen(num);
  char *res = (char*)malloc(sizeof(char*) * (length*2));
  res[0] = '\0';

  char tmp_res[3];

  while (1) {
    next_index = index + 1;
    if (next_index >= length) {
      sprintf(tmp_res, "%d%c", count, num[index]);
      strcat(res, tmp_res);
      break;
    } else {
      if (num[index] == num[next_index]) {
        count++;
      } else {
        sprintf(tmp_res, "%d%c", count, num[index]);
        strcat(res, tmp_res);
        count = 1;
      }
    }
    index++;
  }

  return res;
}


char* countAndSay(int n) {
  char *res = (char*)malloc(sizeof(char*) * MAX_LEN);
  res[0] = '1';
  res[1] = '\0';
  int i;
  for (i = 1; i < n; i++) {
    res = next_num(res);
  }
  return res;
}

File: leetcode/38-Count-And-Say/answer.py
# https://leetcode.com/problems/count-and-say/

def get_num(start):

    index = 0
    count = 1
    length = len(start)

    temp_list = []

    while True:
        next_one = index + 1
        if next_one >= length:
            temp = str(count) + start[index]
            break
        else:
            if start[index] == start[next_one]:
                count += 1
            else:
                temp = str(count) + start[index]
                temp_list.append(temp)
                count = 1

        index += 1
    
    temp_list.append(temp)
    return ''.join(temp_list)


class Solution(object):
    def countAndSay(self, n):
        """
        :type n: int
        :rtype: str
        """
        
        start = '1'
        for i in range(1, n):
            start = get_num(start)
        return start
